libopenage/renderer/stages/world/shader_template.cpp

```diff
@@ -0,0 +1,65 @@
+// Copyright 2024-2025 the openage authors. See copying.md for legal info.
+
+#include "shader_template.h"
+
+#include <cstring>
+
+#include "error/error.h"
+#include "log/log.h"
+
+namespace openage::renderer::world {
+
+ShaderTemplate::ShaderTemplate(const util::Path &template_path) {
+	auto file = template_path.open();
+	this->template_code = file.read();
+	file.close();
+}
+
+void ShaderTemplate::load_snippets(const util::Path &snippet_path) {
+	// load config here
+	util::Path snippet_path_copy = snippet_path;
+	for (const auto &entry : snippet_path_copy.iterdir()) {
+		if (entry.get_name().ends_with(".snippet")) {
+			add_snippet(entry);
+		}
+	}
+}
+
+void ShaderTemplate::add_snippet(const util::Path &snippet_path) {
+	std::string file_name = snippet_path.get_name();
+	std::string name = file_name.substr(0, file_name.find_last_of('.'));
+	auto file = snippet_path.open();
+	std::string content = file.read();
+	file.close();
+
+	snippets[name] = content;
+}
+
+std::string ShaderTemplate::generate_source() const {
+	std::string result = template_code;
+
+	// Process each placeholder
+	for (const auto &[name, snippet_code] : snippets) {
+		std::string placeholder = "// PLACEHOLDER: " + name;
+		size_t pos = result.find(placeholder);
```

Comment:
Idea for a faster approach: Instead of searching the whole source string for each snippet, we could search for the placeholder positions during initialization of the class and save the positions. Then when we want to insert the snippets later, we just have to look up the positions again.

Probably not that big of a deal right now, but it might be more performant on large template files.